src/range_proof.rs

```diff
@@ -1,18 +1,21 @@
 use bulletproofs::r1cs::{ConstraintSystem, LinearCombination, R1CSError};
 use curve25519_dalek::scalar::Scalar;
 
+use crate::value::SignedInteger;
+
 /// Enforces that the quantity of v is in the range [0, 2^n).
 pub fn range_proof<CS: ConstraintSystem>(
     cs: &mut CS,
     mut v: LinearCombination,
-    v_assignment: Option<u64>,
+    v_assignment: Option<SignedInteger>,
     n: usize,
 ) -> Result<(), R1CSError> {
     let mut exp_2 = Scalar::one();
     for i in 0..n {
         // Create low-level variables and add them to constraints
         let (a, b, o) = cs.allocate(|| {
-            let q: u64 = v_assignment.ok_or(R1CSError::MissingAssignment)?;
+            // let q: u64 = v_assignment.ok_or(R1CSError::MissingAssignment)?;
+            let q: u64 = v_assignment.ok_or(R1CSError::MissingAssignment)?.abs();
```

Comment:
lets have `SignedInteger.to_u64() -> Option<u64>` that would fail (`None`) if it's negative. Then the prover has to fail like so:

```rust
let q: u64 = v_assignment.ok_or(R1CSError::MissingAssignment)?.to_u64().ok_or(R1CSError::InvalidAssignment)?;
```

Reply:
together with i128 conversion (see below), we wouldn't need `abs()` and `sign()`, which means we'd treat Negative/Positive variants as completely non-interoperable as distinct types. (Meaning, there's no reason to even have unifying API like `abs()`.)